File: rust/459.repeated-substring-pattern.rs
/*
 * @lc app=leetcode id=459 lang=rust
 *
 * [459] Repeated Substring Pattern
 */

// @lc code=start
impl Solution {
    pub fn repeated_substring_pattern(s: String) -> bool {
        let len = s.len();

        // Solution 1:
        // for i in 1..=len / 2 {
        //     if len % i == 0 && s[..i].repeat(len / i) == s {
        //         return true;
        //     }
        // }
        // false

        // Solution 2:
        s.repeat(2)[1..len * 2 - 1].contains(&s)
    }
}
// @lc code=end

File: javascript/459.repeated-substring-pattern.js
/*
 * @lc app=leetcode id=459 lang=javascript
 *
 * [459] Repeated Substring Pattern
 */

// @lc code=start
/**
 * @param {string} s
 * @return {boolean}
 */
var repeatedSubstringPattern = function (s) {
  // Solution 1:
  // const l = s.length;
  // for (let i = 1; i <= l >> 1; i++) {
  //   if (l % i === 0 && s.slice(0, i).repeat(l / i) === s) return true;
  // }
  // return false;

  // Solution 2:
  return s.repeat(2).slice(1, -1).includes(s);
};
// @lc code=end
